### DISCUSSION TITLE
Asyncio issues

### DISCUSSION THREAD

**Discussion by francijb** [hammem/monarchmoney#101] [Q&A] [answered]
I am not super familiar with asyncio. I am building a program to build custom views on the monarch data. However, I am running into an issue whenever with the asyncio part:

mm = MonarchMoney()
await mm.login(user_name, pw)

SyntaxError: 'await' outside async function

When run individually, they work fine. But when trying to run them all at once, it triggers an error. Is there a way around this?

**Comment by hammem** [comment, marked as answer]
@francijb , can you share a bit more of the code?  I have a feeling you are calling the `await` inside of a non-async function, but would be good to see the code around it to confirm.

As an example, this will cause the error you are seeing:

```python
def my_func() {
  ...
  mm = MonarchMoney()
  await mm.login(user_name, pw) # This fails because an async keyword is being used inside a non-async function
}
```

If that's what's happening, you have two options:

  - convert the outer function to be `async def my_func()` 
    - That will mean you need call it with `await` as well
    - This may not be possible, if you don't control the API of the outer function
  - instead of using `await` use the asyncio library functions to wrap it in a non-async helper

For the first option, this is what your refactored code would look like

```python
from monarchmoney import MonarchMoney

async def my_func() {
    mm = MonarchMoney()
    await mm.login(user_name, pw) # everything is async now, so it works!
}
```

For the latter, you can do the following:

```python
import asyncio
from monarchmoney import MonarchMoney

def my_func() {
    mm = MonarchMoney()
    asyncio.get_event_loop().run_until_complete(mm.login(user_name, pw)) # 'converts' the async function in to a synchronous one
}
```

**Reply by francijb** [reply]
Thank you so much. You got it exactly. I was creating a function like you described. The `async def myfunc()` solved the problem. I appreciate it.
